src/waltham/waltham-connection.c

```diff
@@ -420,3 +410,136 @@ wth_connection_assert_side(struct wth_connection *conn,
+{
+	fprintf(stderr, "wth_display.delete_id(%d)\n", id);
+}
+
+static void
+display_server_version(struct wth_display *d, uint32_t ver)
+{
+	fprintf(stderr, "wth_display.server_version(%d)\n", ver);
+}
+
+static const struct wth_display_listener display_listener = {
+	display_error,
+	display_delete_id,
+	display_server_version
+};
+
+WTH_EXPORT struct wthp_registry *
+wth_connection_get_registry(struct wth_connection *conn)
+{
+	ASSERT_CLIENT_SIDE(conn);
+
+	wth_display_set_listener(conn->display, &display_listener, conn);
+	return wth_display_get_registry(conn->display);
+}
+
+WTH_EXPORT struct wthp_callback *
+wth_connection_sync(struct wth_connection *conn)
+{
+	ASSERT_CLIENT_SIDE(conn);
+
+	return wth_display_sync(conn->display);
+}
+
+/* END wthp_display client implementation */
+
+/* BEGIN wthp_display server implementation */
+
+/* TODO: these three declaration are copied from waltham-server.h, which can’t
+ * be included together with waltham-client.h.  Find a way to make it possible
+ * to include them both, or split this server implementation in another file.
+ */
+
+struct wth_display_interface {
+	void (*client_version) (struct wth_display * wth_display, uint32_t client_version);
+	void (*sync) (struct wth_display * wth_display, struct wthp_callback * callback);
+	void (*get_registry) (struct wth_display * wth_display, struct wthp_registry * registry);
+};
+
+void
+wthp_callback_send_done (struct wthp_callback * wthp_callback, uint32_t callback_data);
+
+void
+wth_display_send_error (struct wth_display * wth_display, struct wth_object * object_id, uint32_t code, const char * message);
+
+static void
+display_handle_client_version(struct wth_display *wth_display,
+			      uint32_t client_version)
+{
+	wth_object_post_error((struct wth_object *)wth_display, 0,
+			      "unimplemented: %s", __func__);
+}
+
+static void
+display_handle_sync(struct wth_display *wth_display,
+                    struct wthp_callback *callback)
+{
+	wth_debug("Client requested wth_display.sync");
+	wthp_callback_send_done(callback, 0);
+	wthp_callback_free(callback);
+}
+
+static void
+display_handle_get_registry(struct wth_display *wth_display,
+			    struct wthp_registry *registry)
+{
+	struct wth_object *disp_object = (struct wth_object *)wth_display;
+	void *user_pointer = wth_object_get_user_data(disp_object);
+	disp_object->connection->registry_callback(registry, user_pointer);
+}
+
+static const struct wth_display_interface display_implementation = {
+	display_handle_client_version,
+	display_handle_sync,
+	display_handle_get_registry
+};
+
+WTH_EXPORT void
+wth_connection_set_registry_callback(struct wth_connection *conn,
+                                     void (*callback)(struct wthp_registry *, void *),
+                                     void *user_pointer)
+{
+	ASSERT_SERVER_SIDE(conn);
+
+	wth_object_set_listener((struct wth_object *)conn->display, (void (**)(void))&display_implementation, user_pointer);
```

Comment:
Does this mean I should add another parameter to `wth_connection_from_fd()` and `wth_accept()` for the `user_pointer`?  I don’t like this solution much. :/

Reply:
No. You should set up the vfunc table in `wth_connection_from_fd()`, but set the user pointer in `wth_connection_set_registry_callback()` as it is.
If this means you can no longer use the `wth_display` user data slot for it, then so be it. Semantically the `wth_display` user data slot belongs to the `wth_display` implementation which is internal to libwaltham, but the slot may have no practical use since libwaltham stashes everything in `wth_connection`.
I think it's perfectly fine to just add a `void *registry_cb_user_data` in `struct wth_connection`, you already have the `registry_callback` member in there and the two belong together. Using the `wth_display` user data may have been a little too clever.